File: Data Structures & Algorithms/find-duplicate-integer/submission-1.py
class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        # 1. 把array看成linked list, index看為值 value看為指針
        # 2. Floyd's 演算法

        slow = fast = 0
        while True:
            slow = nums[slow]
            fast = nums[nums[fast]]
            if slow == fast:
                break
        
        slow2 = 0
        while True:
            slow = nums[slow]
            slow2 = nums[slow2]
            if slow == slow2:
                return slow

File: Data Structures & Algorithms/reorder-linked-list/submission-1.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} head
     * @return {void}
     */
    reorderList(head) {

        // 分割成兩個 Linked List
        let slow = head;
        let fast = head;
        while (fast.next && fast.next.next) {
            slow = slow.next;
            fast = fast.next.next;
        }
        
        // 將後半部分的 next 反轉 ( "->" 變成 "<-" )
        let curr = slow.next;
        slow.next = null;
        let prev = null;
        while (curr) {
            let nextNode = curr.next;
            curr.next = prev;
            prev = curr;
            curr = nextNode;
        }

        // 將前後段依序組合
        let p1 = head;
        let p2 = prev;
        while(p1 && p2) {
            let nextP1 = p1.next;
            let nextP2 = p2.next;
            p1.next = p2;
            p2.next = nextP1;
            p1 = nextP1;
            p2 = nextP2;
        }
    }
}
